File: src/fuzzy/rule/antecedent/factory/heuristic_antecedent_factory.py
from main.consts import Consts
from src.fuzzy.rule.antecedent.factory.abstract_antecedent_factory import AbstractAntecedentFactory
from src.fuzzy.knowledge.knowledge import Knowledge
from src.fuzzy.rule.antecedent.antecedent import Antecedent
import numpy as np
import random


class HeuristicAntecedentFactory(AbstractAntecedentFactory):
    __dimension = None
    __training_set = None

    def __init__(self, training_set):
        self.__dimension = Knowledge.get_instance().get_num_dim()
        self.__training_set = training_set

    def select_antecedent_part(self, index):
        pattern = self.__training_set.get_pattern(index)
        return self.calculate_antecedent_part(pattern)

    def calculate_antecedent_part(self, pattern):
        attribute_array = pattern.get_attributes_vector()

        if Consts.IS_DONT_CARE_PROBABILITY:
            dc_rate = Consts.DONT_CARE_RT
        else:
            dc_rate = max((self.__dimension - Consts.ANTECEDENT_NUMBER_NOT_DONT_CARE) / self.__dimension, Consts.DONT_CARE_RT)

        antecedent_indices = np.zeros(self.__dimension, dtype=np.int_)
        knowledge = Knowledge.get_instance()

        for dim_i in range(self.__dimension):
            # DC
            if random.random() < dc_rate:
                antecedent_indices[dim_i] = 0  # The first fuzzy set is don't care
                continue

            # Categorical judge
            # TODO: Check back later if it's used in the code
            if attribute_array[dim_i] < 0:
                antecedent_indices[dim_i] = int(attribute_array[dim_i])
                continue

            # Numerical (get a random fuzzy set index using the membership value)
            num_fuzzy_sets_not_dc = knowledge.get_num_fuzzy_sets(dim_i)-1
            if num_fuzzy_sets_not_dc < 1:
                antecedent_indices[dim_i] = 0  # don't care
                continue

            mb_values_inc_sums = np.zeros(num_fuzzy_sets_not_dc, dtype=np.float_)
            sum_mb_values = 0
            for h in range(num_fuzzy_sets_not_dc):
                sum_mb_values += knowledge.get_membership_value(attribute_array[dim_i], dim_i, h+1)
                mb_values_inc_sums[h] = sum_mb_values

            arrow = random.random() * sum_mb_values

            for h in range(num_fuzzy_sets_not_dc):
                if arrow < mb_values_inc_sums[h]:
                    antecedent_indices[dim_i] = h+1
                    break

        return antecedent_indices

    def create(self, num_rules=None):
        indices = self.create_antecedent_indices(num_rules)

        antecedent_objects = np.array([Antecedent(self.select_antecedent_part((indices[i]))) for i in range(num_rules)], dtype=object)

        if num_rules == 1:
            antecedent_objects = antecedent_objects[0]
        return antecedent_objects

    def create_antecedent_indices(self, num_rules=None):
        data_size = self.__training_set.get_size()
        if num_rules is None:
            pattern_index = random.randint(0, data_size - 1)
            return self.select_antecedent_part(pattern_index)

        if num_rules <= self.__training_set.get_size():
            return np.random.choice(list(range(self.__training_set.get_size())), num_rules, replace=False)
        else:
            indices = [i for i in range(data_size)] * (num_rules // data_size)

            num_remaining_indices = num_rules % data_size
            remaining_indices = np.random.choice(list(range(self.__training_set.get_size())), num_remaining_indices,
                                                 replace=False)
            return np.concatenate((indices, remaining_indices))

    def __str__(self):
        return "HeuristicAntecedentFactory [dimension=" + str(self.__dimension) + "]"

    def __copy__(self):
        return HeuristicAntecedentFactory(self.__training_set)

File: src/mofgbmlpy/core/fuzzy/rule/antecedent/factory/heuristic_antecedent_factory.cpp
#include "heuristic_antecedent_factory.hpp"
#include <memory>
#include <stdexcept>
#include <algorithm>
#include <numeric>

HeuristicAntecedentFactory::HeuristicAntecedentFactory(const std::shared_ptr<Dataset>& training_set,
                                                       const std::shared_ptr<Knowledge>& knowledge,
                                                       bool is_dc_probability,
                                                       double dc_rate,
                                                       int antecedent_number_do_not_dont_care,
                                                       const std::shared_ptr<RandomGenerator>& random_generator)
    : AbstractAntecedentFactory(knowledge, random_generator),
      training_set(training_set),
      is_dc_probability(is_dc_probability),
      dc_rate(dc_rate),
      antecedent_number_do_not_dont_care(antecedent_number_do_not_dont_care)
{
    if (!knowledge || knowledge->get_num_dim() == 0) {
        throw std::runtime_error("Knowledge is uninitialized or empty");
    }

    if (!training_set || training_set->get_size() == 0) {
        throw std::runtime_error("Training set is uninitialized or empty");
    }

    if (knowledge->get_num_dim() != training_set->get_num_dim()) {
        throw std::invalid_argument("Knowledge and training set dimensions do not match");
    }

    if (dc_rate < 0.0 || dc_rate > 1.0) {
        throw std::invalid_argument("dc_rate must be in the range [0.0, 1.0]");
    }

    if (antecedent_number_do_not_dont_care < 0) {
        throw std::invalid_argument("antecedent_number_do_not_dont_care must be non-negative");
    }

    if (!is_dc_probability)
        this->dc_rate = std::max(
            (knowledge->get_num_dim() - antecedent_number_do_not_dont_care) / static_cast<double>(knowledge->
                get_num_dim()), dc_rate);
}

HeuristicAntecedentFactory::HeuristicAntecedentFactory(const HeuristicAntecedentFactory& other)
    : HeuristicAntecedentFactory(other.training_set, other.knowledge,
                                 other.is_dc_probability, other.dc_rate,
                                 other.antecedent_number_do_not_dont_care,
                                 other.random_generator) {}

std::vector<int> HeuristicAntecedentFactory::select_antecedent_part(int index) const {
    auto pattern = training_set->get_pattern(index);
    return calculate_antecedent_part(*pattern);
}

std::vector<int> HeuristicAntecedentFactory::calculate_antecedent_part(const Pattern& pattern) const {
    if (pattern.get_num_dim() != knowledge->get_num_dim()) {
        throw std::invalid_argument("Pattern dimension mismatch");
    }

    int dim = knowledge->get_num_dim();
    std::vector<int> antecedent_indices(dim, 0);

    for (int i = 0; i < dim; ++i) {
        if (random_generator->random() < dc_rate) {
            // DC
            antecedent_indices[i] = 0;  // The first fuzzy set (index = 0) is don't care
            continue;
        }

        // Categorical judge
        double attr_val = pattern.get_attribute_value(i);
        if (attr_val < 0) {
            antecedent_indices[i] = static_cast<int>(attr_val);
            continue;
        }

        // Numerical (get a random fuzzy set index using the membership value)
        int num_fuzzy_sets = knowledge->get_num_fuzzy_sets(i) - 1;
        if (num_fuzzy_sets < 1) {
            antecedent_indices[i] = 0;  // don't care
            continue;
        }

        std::vector<double> mb_values_inc_sum(num_fuzzy_sets, 0.0);
        double sum_mb = 0.0;
        for (int h = 0; h < num_fuzzy_sets; ++h) {
            sum_mb += knowledge->get_membership_value(attr_val, i, h + 1);
            mb_values_inc_sum[h] = sum_mb;
        }

        double arrow = random_generator->random() * sum_mb;

        for (int h = 0; h < num_fuzzy_sets; ++h) {
            if (arrow < mb_values_inc_sum[h]) {
                antecedent_indices[i] = h + 1;
                break;
            }
        }
    }

    return antecedent_indices;
}

std::vector<std::shared_ptr<Antecedent>> HeuristicAntecedentFactory::create(int num_rules) const {
    if (num_rules <= 0) {
        throw std::invalid_argument("num_rules must be positive");
    }

    std::vector<std::vector<int>> indices = create_antecedent_indices(num_rules);
    std::vector<std::shared_ptr<Antecedent>> antecedents(indices.size());
    for (size_t i = 0; i < indices.size(); ++i) {
        antecedents[i] = std::make_shared<Antecedent>(indices[i], knowledge);
    }

    return antecedents;
}

std::vector<std::vector<int>> HeuristicAntecedentFactory::create_antecedent_indices_from_pattern(const Pattern& pattern) const {
    return { calculate_antecedent_part(pattern) };
}

std::vector<std::vector<int>> HeuristicAntecedentFactory::create_antecedent_indices(int num_rules) const {
    if (num_rules <= 0) {
        throw std::invalid_argument("num_rules must be positive");
    }

    int data_size = training_set->get_size();

    if (num_rules == 1) {
        return { select_antecedent_part(random_generator->randint(0, data_size)) };
    }

    std::vector<int> pattern_indices;
    pattern_indices.reserve(num_rules);

    if (num_rules <= data_size) {
        pattern_indices = random_generator->generate_integers(num_rules, data_size, false);
    } else {
        for (int i = 0; i < num_rules / data_size; i++) {
            for (int j = 0; j < data_size; j++) {
                pattern_indices.push_back(j);
            }
        }
        int num_remaining_indices = num_rules % data_size;

        std::vector<int> remaining_indices = random_generator->generate_integers(num_remaining_indices, data_size, false);

        for (int idx : remaining_indices) {
            pattern_indices.push_back(idx);
        }
    }

    std::vector<std::vector<int>> new_antecedent_indices;
    new_antecedent_indices.reserve(num_rules);

    for (int idx : pattern_indices) {
        new_antecedent_indices.push_back(select_antecedent_part(idx));
    }
    return new_antecedent_indices;
}

std::shared_ptr<Dataset> HeuristicAntecedentFactory::get_training_set() const {
    return training_set;
}

bool HeuristicAntecedentFactory::get_is_dc_probability() const {
    return is_dc_probability;
}

double HeuristicAntecedentFactory::get_dc_rate() const {
    return dc_rate;
}

int HeuristicAntecedentFactory::get_antecedent_number_do_not_dont_care() const {
    return antecedent_number_do_not_dont_care;
}

bool HeuristicAntecedentFactory::operator==(const HeuristicAntecedentFactory& other) const {
    return (training_set == other.get_training_set() &&
            *knowledge == *other.get_knowledge() &&
            is_dc_probability == other.get_is_dc_probability() &&
            dc_rate == other.get_dc_rate() &&
            antecedent_number_do_not_dont_care == other.get_antecedent_number_do_not_dont_care());
}

AbstractAntecedentFactory* HeuristicAntecedentFactory::clone() const {
    return new HeuristicAntecedentFactory(*this);
}

HeuristicAntecedentFactory::operator std::string() const
{
    return "HeuristicAntecedentFactory [training_set_size=" + std::to_string(training_set->get_size()) +
           ", dimension=" + std::to_string(knowledge->get_num_dim()) + "]";
}

